lib/Dialect/Arc/Transforms/MergeIfs.cpp

```diff
@@ -0,0 +1,365 @@
+      // `insertionPoint` lookup, but can cause significant linear scanning to
+      // find the op before which we want to insert.
+      auto &insertionPoint = insertionPoints[earliest.op];
+      if (insertionPoint) {
+        auto order = opOrder.lookup(insertionPoint);
+        assert(order.first == earliest.order.first);
+        assert(order.second >= earliest.order.second);
+        earliest.op = insertionPoint;
+        earliest.order = order;
+      }
+      while (auto *prevOp = earliest.op->getPrevNode()) {
+        auto order = opOrder.lookup(prevOp);
+        if (order.first != earliest.order.first)
+          break;
+        assert(order.second > earliest.order.second);
+        earliest.op = prevOp;
+        earliest.order = order;
+      }
+      insertionPoint = earliest.op;
+
+      // Only move if the op isn't already in the right spot.
+      if (op.getNextNode() != earliest.op) {
+        LLVM_DEBUG(llvm::dbgs() << "- Moved " << op << "\n");
+        op.moveBefore(earliest.op);
+        ++pass.numOpsMovedToUser;
+        anyChanges = true;
+      }
+
+      // Update the current op's order to reflect where it has been inserted.
+      // This ensures that later moves to the same pile of moved ops do not
+      // reorder the operations.
+      order = earliest.order;
+      assert(order.second < unsigned(-1));
+      ++order.second;
+      opOrder[&op] = order;
+    }
+  }
+}
+
+void Sinker::mergeIfs() {
+  DenseSet<Value> writes, reads;
+
+  scf::IfOp lastOp;
+  for (auto ifOp : rootBlock.getOps<scf::IfOp>()) {
+    auto prevIfOp = std::exchange(lastOp, ifOp);
+    if (!prevIfOp)
+      continue;
+
+    // Only handle simple cases for now. (Same condition, no results, only
+    // single blocks, and both ifs either have or don't have an else region.)
+    if (ifOp.getCondition() != prevIfOp.getCondition())
+      continue;
+    if (ifOp.getNumResults() != 0 || prevIfOp.getNumResults() != 0)
+      continue;
+    if (!ifOp.getThenRegion().hasOneBlock() ||
+        !prevIfOp.getThenRegion().hasOneBlock())
+      continue;
+    if (ifOp.getElseRegion().empty() != prevIfOp.getElseRegion().empty())
+      continue;
+    if (!ifOp.getElseRegion().empty() &&
+        (!ifOp.getElseRegion().hasOneBlock() ||
+         !prevIfOp.getElseRegion().hasOneBlock()))
+      continue;
+
+    // Try to move ops in between the if ops above the previous if in order to
+    // make them immediately adjacent.
+    if (ifOp->getPrevNode() != prevIfOp) {
+      // Determine the side effects inside the previous if op.
+      bool hasSideEffects = false;
+      writes.clear();
+      reads.clear();
+      prevIfOp.walk([&](Operation *op) {
+        if (auto ptr = getPointerWrittenByOp(op))
+          writes.insert(ptr);
+        else if (auto ptr = getPointerReadByOp(op))
+          reads.insert(ptr);
+        else if (!hasSideEffects &&
+                 !op->hasTrait<OpTrait::HasRecursiveMemoryEffects>() &&
```

Comment:
Same as above, why do we need the check for the trait if we already have `isMemoryEffectFree`?